CIP-builtin-value/README.md

```diff
@@ -0,0 +1,138 @@
+     res2 = builtinMapValues (builtinMapFilterKeys (builtinEquals 0)) res
+ in builtinMapFilterKeys ((/=) BI.null) res2
+```
+
+This requires a non-trivial number of limited higher order builtins, and is not a very intuitive function to define. Language designers targeting Plutus Core (Aiken, Plutarch, Plu-Ts, Scalus, Purus, Helios) will still need to implement the value operations using the `ByteStringMap` in their standard library which means that on top of being less performant than executing natively, there is still security risk that some implementation is insecure. On top of all that, the above implementation is non-canonical! This means that different language standard libraries could implement `unionValue` differently (for instance not filtering 0 entries or empty inner map entries) and then implement `equalsValue` to handle equality regardless of zero entries or empty map entries (ex. Plinth currently does this). 
+ 
+We argue that it should not be the responsibility of language authors to implement secure operations over a type that is primitive to the domain task.
+
+### MaryEraValue 
+
+A _map_ is a collection of _values_ indexed by _keys_. We consider a map to be _well-defined_ if each key uniquely identifies each value. The canonical representation of Mary-era multi-asset Values in PlutusCore is a nested map where the keys of the outer map are `AsData CurrencySymbol`s and the keys of the inner map are `AsData TokenName`s and the _values_ of the inner map are `AsData Integer` (representing the quantity of the asset). 
+
+Here are the important invariants of the PlutusCore representation of the Mary-era Value types from the Cardano ledger:
+1. All _keys_ in the outer map are unique (each `CurrencySymbol` appears only once)
+2. All _keys_ in the inner map are unique (each `TokenName` appears only once)
+3. The _keys_ in the outer map are ordered lexographically (`CurrencySymbol`s appear in ascending order)
+4. The _keys_ in the inner map are ordered lexographically (`TokenName`s appear in ascending order)
+5. There are no empty inner maps (The _value_ associated with any given `CurrencySymbol` _key_ cannot be an empty map)
+6. There are no zero-quantity _values_ in the inner maps (The _value_ associated with any given `TokenName` _key_ in an inner-map cannot be zero).
+
+### Mary-era Value as a builtin type
+
+The only remaining solutions are to consider:
+
+1. Introduce builtin operations for efficient manipulation of `BuiltinData = Value` that take into consideration the invariants of ledger multi-asset values.
+2. Introduce a new builtin type `BuiltinMaryEraValue` with efficient Value operations that take into consideration the invariants of ledger multi-asset values. 
+
+
+## Specification
+
+We propose the addition of the following builtin functions: 
+
+All functions below work with:
+```
+type BuiltinMaryEraValue = BuiltinData -- or possibly (BuiltinList BuiltinCurrencySymbol (BuiltinMap BuiltinTokenName BuiltinData)) 
```

Comment:
See above, we don't think that either representation (`BuiltinData`/`BuiltinList`) is suitable and this CIP needs to introduce `BuiltinMaryEraValue` as a new builtin type.